File: pages/refactorings/replace-mutable-fields-with-lenses/02-rename-tree/Spec.hs
{-# LANGUAGE OverloadedStrings #-}
module Main where

import Control.Monad (unless)
import System.Exit (exitFailure)
import Data.Char (isUpper)
import Hedgehog
import qualified Hedgehog.Gen as Gen
import qualified Hedgehog.Range as Range
import qualified Before
import qualified After

-- A neutral tree, so one generated input feeds both Expr types.
data T = TVar String Int | TApp T T | TLam String T
  deriving (Eq, Show)

genName :: Gen String
genName = Gen.string (Range.linear 0 4) Gen.alpha

genT :: Int -> Gen T
genT depth =
  let leaf = TVar <$> genName <*> Gen.int (Range.linear (-10) 10)
  in if depth == 0 then leaf
     else Gen.choice
       [ leaf
       , TApp <$> genT (depth - 1) <*> genT (depth - 1)
       , TLam <$> genName <*> genT (depth - 1)
       ]

toBefore :: T -> Before.Expr
toBefore (TVar n r) = Before.EVar (Before.Var n r)
toBefore (TApp f x) = Before.EApp (toBefore f) (toBefore x)
toBefore (TLam b e) = Before.ELam b (toBefore e)

fromBefore :: Before.Expr -> T
fromBefore (Before.EVar v)  = TVar (Before.vName v) (Before.vRef v)
fromBefore (Before.EApp f x) = TApp (fromBefore f) (fromBefore x)
fromBefore (Before.ELam b e) = TLam b (fromBefore e)

toAfter :: T -> After.Expr
toAfter (TVar n r) = After.EVar (After.Var n r)
toAfter (TApp f x) = After.EApp (toAfter f) (toAfter x)
toAfter (TLam b e) = After.ELam b (toAfter e)

fromAfter :: After.Expr -> T
fromAfter (After.EVar v)  = TVar (After.vName v) (After.vRef v)
fromAfter (After.EApp f x) = TApp (fromAfter f) (fromAfter x)
fromAfter (After.ELam b e) = TLam b (fromAfter e)

prop_agrees :: Property
prop_agrees = property $ do
  t <- forAll (genT 4)
  fromBefore (Before.renameAll (toBefore t))
    === fromAfter (After.renameAll (toAfter t))

allNamesUpper :: T -> Bool
allNamesUpper (TVar n _) = all isUpper n
allNamesUpper (TApp f x) = allNamesUpper f && allNamesUpper x
allNamesUpper (TLam _ b) = allNamesUpper b

refsUnchanged :: T -> T -> Bool
refsUnchanged (TVar _ r1) (TVar _ r2) = r1 == r2
refsUnchanged (TApp f1 x1) (TApp f2 x2) =
  refsUnchanged f1 f2 && refsUnchanged x1 x2
refsUnchanged (TLam _ b1) (TLam _ b2) = refsUnchanged b1 b2
refsUnchanged _ _ = False

prop_all_upper :: Property
prop_all_upper = property $ do
  t <- forAll (genT 4)
  let renamed = fromAfter (After.renameAll (toAfter t))
  allNamesUpper renamed === True
  refsUnchanged renamed t === True

main :: IO ()
main = do
  ok <- checkParallel $ Group "Props"
    [ ("renameAll: Before == After", prop_agrees)
    , ("every Var name is uppercased, refs unchanged", prop_all_upper)
    ]
  unless ok exitFailure

File: pages/refactorings/replace-mutable-fields-with-lenses/02-rename-tree/Spec.scala
//> using scala 3.3.4
//> using dep qa.hedgehog::hedgehog-core:0.14.0
//> using dep qa.hedgehog::hedgehog-runner:0.14.0
import hedgehog.*, hedgehog.core.*, hedgehog.runner.*

object Props extends Properties:
  def tests: List[Test] = List(
    property("renameAll: Before == After", agrees),
    property("every Var name is uppercased after the walk", allUpper),
  )

  // A neutral tree, so one generated input feeds both Expr types.
  enum T:
    case TVar(name: String, ref: Int)
    case TApp(f: T, x: T)
    case TLam(bind: String, body: T)

  val genName: Gen[String] =
    Gen.alpha.list(Range.linear(0, 4)).map(_.mkString)
  def genT(depth: Int): Gen[T] =
    val leaf =
      for n <- genName; r <- Gen.int(Range.linear(-10, 10))
      yield T.TVar(n, r)
    if depth == 0 then leaf
    else
      Gen.choice1(
        leaf,
        for
          f <- genT(depth - 1)
          x <- genT(depth - 1)
        yield T.TApp(f, x),
        for b <- genName; body <- genT(depth - 1) yield T.TLam(b, body),
      )

  def toBefore(t: T): Before.Expr = t match
    case T.TVar(n, r) => Before.Expr.EVar(Before.Var(n, r))
    case T.TApp(f, x) => Before.Expr.EApp(toBefore(f), toBefore(x))
    case T.TLam(b, e) => Before.Expr.ELam(b, toBefore(e))
  def fromBefore(e: Before.Expr): T = e match
    case Before.Expr.EVar(v)     => T.TVar(v.name, v.ref)
    case Before.Expr.EApp(f, x)  => T.TApp(fromBefore(f), fromBefore(x))
    case Before.Expr.ELam(b, e)  => T.TLam(b, fromBefore(e))
  def toAfter(t: T): After.Expr = t match
    case T.TVar(n, r) => After.Expr.EVar(After.Var(n, r))
    case T.TApp(f, x) => After.Expr.EApp(toAfter(f), toAfter(x))
    case T.TLam(b, e) => After.Expr.ELam(b, toAfter(e))
  def fromAfter(e: After.Expr): T = e match
    case After.Expr.EVar(v)     => T.TVar(v.name, v.ref)
    case After.Expr.EApp(f, x)  => T.TApp(fromAfter(f), fromAfter(x))
    case After.Expr.ELam(b, e)  => T.TLam(b, fromAfter(e))

  def agrees: Property =
    for t <- genT(4).forAll
    yield
      fromBefore(Before.renameAll(toBefore(t)))
        ==== fromAfter(After.renameAll(toAfter(t)))

  def allUpper: Property =
    for t <- genT(4).forAll
    yield
      val renamed = fromAfter(After.renameAll(toAfter(t)))
      (allNamesUpper(renamed) && refsUnchanged(renamed, t)) ==== true

  def allNamesUpper(t: T): Boolean = t match
    case T.TVar(n, _)   => n == n.toUpperCase
    case T.TApp(f, x)   => allNamesUpper(f) && allNamesUpper(x)
    case T.TLam(_, b)   => allNamesUpper(b)
  def refsUnchanged(a: T, b: T): Boolean = (a, b) match
    case (T.TVar(_, r1), T.TVar(_, r2)) => r1 == r2
    case (T.TApp(f1, x1), T.TApp(f2, x2)) =>
      refsUnchanged(f1, f2) && refsUnchanged(x1, x2)
    case (T.TLam(_, b1), T.TLam(_, b2)) => refsUnchanged(b1, b2)
    case _ => false

@main def spec(): Unit =
  val results = Props.tests.map { t =>
    val r = Property.check(t.withConfig(PropertyConfig.default),
      t.result, Seed.fromTime())
    println(Test.renderReport(
      "Props", t, r, ansiCodesSupported = false))
    r.status
  }
  if !results.forall(_ == Status.ok) then sys.exit(1)
